File: src/main/scala/com/vc/hard/L25.scala
package com.vc.hard

class ListNode25(var _x: Int = 0) {
 var next: ListNode25 = null
 var x: Int = _x
}

object L25 {
  def reverseKGroup(head: ListNode25, k: Int): ListNode25 = {
    if(head == null || head.next == null) return head

    var n = 0
    var current = head
    while(current != null) {
      n += 1
      current = current.next
    }

    if(k > n) return head

    val dummyNode = new ListNode25()
    dummyNode.next = head
    var prev = dummyNode
    current = prev.next
    var next = current.next

    var i = 0
    while(next != null) {
      i += 1
      if(n < k) return dummyNode.next
      if(i % k == 0) {
         n -= k
         prev = current
         current = current.next
         next = current.next
      }
      else {
        current.next = next.next
        next.next = prev.next
        prev.next = next
        next = current.next
      }
    }
    dummyNode.next
  }
}

File: src/main/scala/com/vc/hard/L25.java
package com.vc.hard;

class ListNode25 {
    int val;
    ListNode25 next;
    ListNode25(int x) { val = x; }
}

class L25 {
    public ListNode25 reverseKGroup(ListNode25 head, int k) {
        if(head == null) return null;

        int n = 0;
        ListNode25 current = head;
        while(current != null) {
            current = current.next;
            n++;
        }

        ListNode25 dummyNode = new ListNode25(-1);
        dummyNode.next = head;

        ListNode25 prev = dummyNode;
        current = prev.next;
        ListNode25 next = current.next;

        int i = 0;
        while(next != null) {
            i++;
            if(n < k) return dummyNode.next;
            if(i % k == 0) {
                n -= k;
                prev = current;
                current = next;
                next = current.next;
            }
            else {
                current.next = next.next;
                next.next = prev.next;
                prev.next = next;
                next = current.next;
            }
        }
        return dummyNode.next;
    }
}
